### PR TITLE
Add an API for managing the settings of Security system indices

### PR DESCRIPTION
**Pull request by AthenaEryma**
This PR adds an API similar to #95342 for managing settings of system indices.

Example calls:

```
GET /_security/settings

PUT /_security/settings
{
    "security": {
        "index.auto_expand_replicas": "0-all"
    },
    "security_tokens": {
            "index.auto_expand_replicas": "0-all"
    },
    "security_profile": {
            "index.auto_expand_replicas": "0-all"
    }
}
```

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java:92]: Wondering about the timeout values here: these will be applied individually to each update request, i.e., the overall composite request may end up running longer than the timeout value, or no?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/settings/UpdateSecuritySettingsAction.java
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/settings/UpdateSecuritySettingsAction.java:138]: Shouldn't we return `existingExceptions` here? Otherwise, `existingExceptions` gets swallowed.

**Review by n1v0lg — COMMENTED** [review]
Nice, I like the API design! 

There's a [small bug around validation](https://github.com/elastic/elasticsearch/pull/97630#discussion_r1262496736) + I think it'd be great to [beef up test coverage](https://github.com/elastic/elasticsearch/pull/97630#discussion_r1262438746) some more since we do have some added complexity here with handling multiple indices in one request.

File: x-pack/docs/en/rest-api/security/update-settings.asciidoc
Comment (gwbrown) [reply x-pack/docs/en/rest-api/security/update-settings.asciidoc:22]: Thanks, saved me some time! I did run the doctests, but before I tweaked the names and introduced a typo while doing so. Oops.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/settings/UpdateSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/settings/UpdateSecuritySettingsAction.java:123]: Good catch, thanks.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/settings/GetSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/settings/GetSecuritySettingsAction.java:29]: It is true, but `allow_restricted_indices` is a bit of a special case - it's a bit of an escape hatch to allow reaching into bits of the system we don't normally permit, rather than something we expect to be commonly used. We definitely don't want to make this rely on an index permission - that's part of the idea of system indices, the typical cluster admin shouldn't have to know about the actual underlying indices as much as possible, and definitely shouldn't have to write configuration around them - in general,we want Elasticsearch features to be managed through APIs that have stability guarantees, rather than interacting directly with system indices which do not.

You have a valid point about this API being under `manage_security` vs `manage` though. This is exactly managing security, but in a way in which it might be permissible for a "normal" cluster admin to do so. I think I could be convinced, I'll think about it while the next round of CI runs.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportGetSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportGetSecuritySettingsAction.java:112]: They're both hacks, but I'm not actually sure why what `TransportUpdateWatcherSettingsAction` does works to prevent the warning, which is emitted from `IndexNameExpressionResolver`, not anything `UpdateSettingsService` is doing.

(Also the way the Watcher does it will break if `.watches` is an alias - which I only caught because security's indices are aliased all the time - so I want to go and make Watcher use this way to resolve it)

Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportGetSecuritySettingsAction.java:112]: Also as I think about it maybe the right way to handle this is to put
```java
    @Override
    public boolean allowSystemIndexAccessByDefault() {
        return true;
    }
```

on the `Rest*SettingsAction` classes. Normally that's not what you want, because you largely should be doing operations on system indices in a system user security context which also prevents the warning, buuuuut I think this might be the oddball case where that's the semantically correct way of expressing what you want?

@williamrandolph do you have an opinion on this one, as the other system indices expert?

Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportGetSecuritySettingsAction.java:112]: Like normally the "correct" way to do this would be to use `OriginSettingClient` and execute transport actions with it to get that system user security context, but that's considerably less efficient than calling `UpdateSettingsService` directly.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java:79]: You're right and IntelliJ even tells me as much. Good catch!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java:118]: Changed this to `createUpdateSettingsRequest` - can help with the length by leaving off the `ForIndex` implicit from the parameters.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java:92]: Because we use the `GroupedActionListener` and fire off the requests rapid-fire in the `.forEach()` just below this, the timeouts should run concurrently. How this works is a lot clearer after the name change to this function you suggested above.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/settings/UpdateSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/settings/UpdateSecuritySettingsAction.java:138]: Uh yep good catch, thanks!

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/security/settings/10_update_security_settings.yml
Comment (gwbrown) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/security/settings/10_update_security_settings.yml:100]: Thanks for pushing for this; I got lazy because we didn't do very thorough testing for the Watcher one.

**Review by n1v0lg — APPROVED** [review]
LGTM. Thanks for iterating!

There are a few smaller things (https://github.com/elastic/elasticsearch/pull/97630#discussion_r1277313603, https://github.com/elastic/elasticsearch/pull/97630#discussion_r1277331051) to address before merging but no need for another review iteration.

As a side note, I played around with this and came across this behavior: https://github.com/elastic/elasticsearch/issues/95347

In short, you can get a node to crash with an assertion if you spin up a cluster with assertions enabled, create a user (or otherwise trigger the security index to be created) then make a request:

```
PUT /_security/settings
{
    "security": {
        "index.number_of_replicas": null
    }
}
```

This trips an assertion as described in the linked issue. I don't think this is something for us to be concerned with since it's just inheriting a bug of the generic index settings API but jotting it down for posterity.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java:116]: Good call, I've added an assertion here to clarify that this ought to be impossible. I've left the `throw` in place, since we still need that for control flow purposes if assertions are disabled and the "impossible" happens.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/settings/UpdateSecuritySettingsActionTests.java
Comment (gwbrown) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/settings/UpdateSecuritySettingsActionTests.java:61]: I don't like relying on randomization to test functionality in unit tests - they're so quick to run, it's better to just run more test cases every time. But I did expand this test to run with one, two, and all three having invalid settings, and I randomized whether the other (not invalid) settings were populated or empty as that shouldn't matter. Does that work for you?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportGetSecuritySettingsAction.java
Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportGetSecuritySettingsAction.java:87]: ... oops. 🤦🏼‍♀️ Thanks

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/SecuritySettingsIT.java
Comment (gwbrown) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/SecuritySettingsIT.java:59]: No, at a minimum when we go to build the new `Metadata` as part of the settings update we make sure all the settings are parseable and/or valid per parsers/validators defined by the setting. For this specific setting, that's [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java#L45-L45). If that fails you should get a `400` (which is the case if I change the value in this test to something not valid, I checked).

File: x-pack/docs/en/rest-api/security/update-settings.asciidoc
Comment (gwbrown) [reply x-pack/docs/en/rest-api/security/update-settings.asciidoc:36]: I'm going to try to make this work, as painful as they are the docs tests are valuable for making sure the examples in the docs actually work.

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/SecuritySettingsIT.java
Comment (n1v0lg) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/SecuritySettingsIT.java:59]: Ah weird -- I thought I got a 200 when I tested this but you're tots right it validates this correctly. Thanks for double-checking!

File: x-pack/docs/en/rest-api/security/update-settings.asciidoc
Comment (gwbrown) [reply x-pack/docs/en/rest-api/security/update-settings.asciidoc:36]: Ugh, I give up. I've spent too much time on this already.

`// TEST[skip:making sure all the indices have been created reliably is difficult]`

We should really figure that out though. It's got to be deterministic.

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/SecuritySettingsIT.java
Comment (gwbrown) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/SecuritySettingsIT.java:50]: I've updated the javadoc here.

**Comment by gwbrown** [comment]
@elasticmachine run elasticsearch-ci/docs

**Comment by garethhumphriesgkc** [comment]
Looking at the commit, it appears this only addresses the `.security*` indices, is that correct?  Is there a plan to support other system indices, e.g.:

- .apm-*
- .ds-*
- .enrich-*
- .fleet-*
- .geoip_databases
- .items-*
- .kibana-*
- .logs-*
- .metrics-*
- .ml-*
- .security-*
- .siem-signals-*
- .tasks
- .transform-*

**Pull request merged**
